pennylane/ops/qubit.py

```diff
@@ -447,6 +454,84 @@ def decomposition(wires):
         return decomp_ops
 
 
+class CU3(Operation):
+    r"""CU3(theta, phi, lambda, wires)
+    The controlled arbitrary single qubit unitary operator.
+
+    .. math::
+
+        CU3(\theta, \phi, \lambda)\ =
+        |0\rangle\langle 0| \otimes I +
+        |1\rangle\langle 1| \otimes U3(\theta,\phi,\lambda) =
+        \begin{bmatrix}
+            1 & 0 & 0 & 0 \\
+            0 & 1 & 0 & 0\\
+            0 & 0 & \cos(\theta/2) & -\exp(i \lambda)\sin(\theta/2)\\
+            0 & 0 & \exp(i \phi)\sin(\theta/2) & \exp(i (\phi + \lambda))\cos(\theta/2)
+        \end{bmatrix}.
+
+    .. note:: The first wire provided corresponds to the **control qubit**.
+
+    **Details:**
+
+    * Number of wires: 2
+    * Number of parameters: 3
+
+    Args:
+        theta (float): polar angle :math:`\theta`
+        phi (float): azimuthal angle :math:`\phi`
+        lambda (float): quantum phase :math:`\lambda`
+        wires (Sequence[int] or int): the subsystem the gate acts on
+    """
+
+    num_params = 3
+    num_wires = 2
+    par_domain = "R"
+    grad_method = "F"
+
+    @classmethod
+    def _matrix(cls, *params):
+        theta, phi, lam = params
+        c = math.cos(theta / 2)
+        s = math.sin(theta / 2)
+
+        return np.array(
+            [
+                [1, 0, 0, 0],
+                [0, 1, 0, 0],
+                [0, 0, c, -s * cmath.exp(1j * lam)],
+                [0, 0, s * cmath.exp(1j * phi), c * cmath.exp(1j * (phi + lam))],
+            ]
+        )
+
+    # Decomposition of CU3 (theta, phi, lam) is as follows
+    # (with wire[0] as control):
+    #
+    # U1((lam + phi) / 2, wires=0)
+    # U1((lam - phi) / 2, wires=1)
+    # CNOT(wires=[0, 1])
+    # U3(-theta / 2, 0, -(phi + lam)/ 2, wires=1)
+    # CNOT(wires=[0, 1])
+    # U3(theta / 2, phi, 0, wires=1)
+    #
+    # U1 and U3 are then further decomposed into Rot and PhaseShift gates.
+    @staticmethod
+    def decomposition(theta, phi, lam, wires):
+        decomp_ops = [
+            PhaseShift((lam + phi) / 2, wires=wires[0]),
+            PhaseShift((lam - phi) / 2, wires=wires[1]),
+            CNOT(wires=wires),
+            Rot(-(phi + lam) / 2, -theta / 2, (phi + lam) / 2, wires=wires[1]),
+            PhaseShift(-(phi + lam) / 2, wires=wires[1]),
+            PhaseShift(0, wires=wires[1]),
+            CNOT(wires=wires),
+            Rot(0, theta / 2, 0, wires=wires[1]),
+            PhaseShift(0, wires=wires[1]),
+            PhaseShift(phi, wires=wires[1]),
```

Comment:
Something that just struck me, is that we don't support adding variables in PennyLane*! So I'm quite surprised the tests pass 🤔 The reason adding variables doesn't work is that we wrap differentiable variables in a `Variable` class. This allows us to track which variables are used in which gate, to automate procedures such as the product rule.

*We _do_ support adding of variables in the new [tape mode](https://pennylane.readthedocs.io/en/latest/code/qml_tape.html), which can be activated by running `qml.enable_tape()`, but this is not default yet.

Reply:
Ah, I see why the tests pass; the operation is not currently being executed on a simulator :)

If you run the following, it should error:

```python
dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev)
def circuit(x, y, z):
    qml.CU3(x, y, z, wires=[0, 1])
    return qml.expval(qml.PauliZ(0))

circuit(0.5, 0.1, 0.2)
```

Reply:
A small addition to trying out the point Josh raised is to enforce the decomposition by having `dev.operations.remove('CU3')`:
```python
import pennylane as qml

dev = qml.device("default.qubit", wires=2)

@qml.qnode(dev)
def circuit(x, y, z):
    qml.CU3(x, y, z, wires=[0, 1])
    return qml.expval(qml.PauliZ(0))

dev.operations.remove('CU3')

circuit(0.5, 0.1, 0.2)
```
```
    519     def decomposition(theta, phi, lam, wires):
    520         decomp_ops = [
--> 521             PhaseShift((lam + phi) / 2, wires=wires[0]),
    522             PhaseShift((lam - phi) / 2, wires=wires[1]),
    523             CNOT(wires=wires),

TypeError: unsupported operand type(s) for +: 'Variable' and 'Variable'
```

Reply:
Ah, I see. To get rid of variable addition, I decomposed the `Rot` and `PhaseShift` gates (which had sum of variables as arguments) further. I used the following facts :

```
Rot (phi, theta, omega) = RZ (omega) * RY (theta) * RZ (phi)

RY (alpha + beta) = RY (alpha) * RY (beta)

RZ (alpha + beta) = RZ (alpha) * RZ (beta)
```

Doing this resulted in the following decomposition, which has no addition of variables (each commented out gate is followed by its decomposition)  : 

```
# PhaseShift((lam + phi) / 2, wires=wires[0]),
PhaseShift(lam / 2, wires=wires[0]),
PhaseShift(phi / 2, wires=wires[0]),
# PhaseShift((lam - phi) / 2, wires=wires[1]),
PhaseShift(lam / 2, wires=wires[1]),
PhaseShift(-phi / 2, wires=wires[1]),
CNOT(wires=wires),
# Rot(-(phi + lam) / 2, -theta / 2, (phi + lam) / 2, wires=wires[1]),
Rot(phi / 2, 0, 0, wires=wires[1]),
Rot(lam / 2, 0, 0, wires=wires[1]),
Rot(0, -theta / 2, 0, wires=wires[1]),
Rot(-phi / 2, 0, 0, wires=wires[1]),
Rot(-lam / 2, 0, 0, wires=wires[1]),
# PhaseShift(-(phi + lam) / 2, wires=wires[1]),
PhaseShift(-phi / 2, wires=wires[1]),
PhaseShift(-lam / 2, wires=wires[1]),
PhaseShift(0, wires=wires[1]),
CNOT(wires=wires),
Rot(0, theta / 2, 0, wires=wires[1]),
PhaseShift(0, wires=wires[1]),
PhaseShift(phi, wires=wires[1]),
```

Note that the `PhaseShift(0)` gates can also be removed.

However, if I do it this way, there are floating point precision errors. On multiplying all these gates together, the floating point imprecision blows up and causes variance of as much as `1e-1`.

To double check my work, I manually checked the decomposition of each gate, which had variable sums as parameters. When multiplying small numbers of gates together, the floating point imprecision is quite reasonable. When the number of gates in the decomposition increases (to 15 in this case), the floating point error becomes significant. 

I will spend some more time trying to figure out what can be done here.

Reply:
Indeed, increasing the tolerance to `1e-1` gives a successful decomposition test run.

Reply:
Here is what the CU3 gate looks like for parameters `theta = 0.65, phi = 0.432, lam = -0.12`.
Ideal expected gate output :
```
[[ 1.        +0.j          0.        +0.j          0.        +0.j
   0.        +0.j        ]
 [ 0.        +0.j          1.        +0.j          0.        +0.j
   0.        +0.j        ]
 [ 0.        +0.j          0.        +0.j          0.94765073+0.j
  -0.31701252+0.03822516j]
 [ 0.        +0.j          0.        +0.j          0.28997395+0.13369072j
   0.90189962+0.29089342j]]

```
If we use the gate decomposition  with 'sums of variables' :
```
[[ 1.00000000e+00+0.00000000e+00j -1.18714181e-17-3.44890007e-18j
   0.00000000e+00+0.00000000e+00j  0.00000000e+00+0.00000000e+00j]
 [-2.73545153e-18-1.26116328e-18j  1.00000000e+00-5.55111512e-17j
   0.00000000e+00+0.00000000e+00j  0.00000000e+00+0.00000000e+00j]
 [ 0.00000000e+00+0.00000000e+00j  0.00000000e+00+0.00000000e+00j
   9.47650726e-01+0.00000000e+00j -3.17012520e-01+3.82251596e-02j]
 [ 0.00000000e+00+0.00000000e+00j  0.00000000e+00+0.00000000e+00j
   2.89973949e-01+1.33690725e-01j  9.01899617e-01+2.90893418e-01j]]
```
If we use the more granular gate decomposition where we eliminate the additions of variables :
```
[[ 0.99873633-0.00803464j  0.01351929+0.04773278j  0.        +0.j
   0.        +0.j        ]
 [-0.01351929+0.04773278j  0.99873633+0.00803464j  0.        +0.j
   0.        +0.j        ]
 [ 0.        +0.j          0.        +0.j          0.9489144 -0.00803464j
  -0.30349323+0.08595794j]
 [ 0.        +0.j          0.        +0.j          0.26245517+0.17496913j
   0.90063595+0.29892806j]]
```

Reply:
1e-1 is a huge difference 😱

Reply:
Yeah, it is very bad. Btw, here's a snippet that shows that all the individual decompositions are correct :

```
theta = 0.65
        phi = 0.432
        lam = -0.12

        op = qml.CU3(theta, phi, lam, wires=[0, 1])
        res = op.decomposition(theta, phi, lam, wires=op.wires)

        assert len(res) == 15

        gate1 = qml.PhaseShift((lam + phi) / 2, wires=0)
        gate2 = qml.PhaseShift((lam - phi) / 2, wires=1)
        gate3 = qml.Rot(-(phi + lam) / 2, -theta / 2, (phi + lam) / 2, wires=1)
        gate4 = qml.PhaseShift(-(phi + lam) / 2, wires=1)

        mats = [e.matrix for e in res]

        subset1 = mats[0:2]
        subset2 = mats[2:4]
        subset3 = mats[5:10]
        subset4 = mats[10:12]

        assert np.allclose(multi_dot(subset1), gate1.matrix, atol=tol, rtol=0)
        assert np.allclose(multi_dot(subset2), gate2.matrix, atol=tol, rtol=0)
        assert np.allclose(multi_dot(subset3), gate3.matrix, atol=tol, rtol=0)
        assert np.allclose(multi_dot(subset4), gate4.matrix, atol=tol, rtol=0)
```

Reply:
Do you think it makes sense to use this snippet instead for the CU3 decomposition test? If so, then maybe we could leave the decomposition this way till tape mode is used by default; at which point we could revert to the old sum-of-variables decomposition. I feel this would maybe make be somewhat okay **for the time being** because :

- It illustrates the point that the decomposition is correct.
- It allows the user to run the gate on the `default.qubit` device while tape mode is not default (though the result will be compromised in accuracy).

Reply:
I found the exact source of the discrepancy. It is to do with how the test runs.
So if we break down a gate which acts on one wire, say a `PhaseShift`, which has sum of two variables as follows :
```
PhaseShift (alpha, beta) = PhaseShift (alpha) * PhaseShift (beta)
```
we will see that upon multiplication we get very little error.

However, the test function multiplies the gates by first moving each constituent `PhaseShift` to the correct Hilbert space using a tensor product :
```
[PhaseShift (alpha) TENSOR Identity]     *     [PhaseShift (beta) TENSOR Identity]
```
Repeated tensor products like this involve multiplication of floating point terms repeatedly with each other, which blows up rapidly. 

I think a fix for this could be to group terms in the decomposition before taking the Tensor Product. For this case, it would be like the following :
```
[PhaseShift (alpha) * PhaseShift (beta)]     TENSOR     Identity
```

Reply:
Here is a working snippet for testing the decomposition in this manner. 

This multiplies all gates together which share the same Hilbert space, and then applies a tensor product based on which wire they're on :

```
theta = 0.65
phi = 0.432
lam = -0.12
tol = 1e-10

op = qml.CU3(theta, phi, lam, wires=[0, 1])
res = op.decomposition(theta, phi, lam, wires=op.wires)

assert len(res) == 15

assert res[0].name == "PhaseShift"
assert res[1].name == "PhaseShift"
assert res[2].name == "PhaseShift"
assert res[3].name == "PhaseShift"
assert res[4].name == "CNOT"
assert res[5].name == "Rot"
assert res[6].name == "Rot"
assert res[7].name == "Rot"
assert res[8].name == "Rot"
assert res[9].name == "Rot"
assert res[10].name == "PhaseShift"
assert res[11].name == "PhaseShift"
assert res[12].name == "CNOT"
assert res[13].name == "Rot"
assert res[14].name == "PhaseShift"

# Grouping gates that are on same wires and have same name
grouped_gates = [[res[0]]]
for i in range(1, len(res)):
    if (
        res[i].wires.tolist() == grouped_gates[-1][0].wires.tolist()
        and res[i].name == grouped_gates[-1][0].name
    ):
        grouped_gates[-1].append(res[i])
    else:
        grouped_gates.append([res[i]])

# Multiplying grouped gates together to yield single gates
grouped_gate_mats = []
for subset in grouped_gates:
    gate_mats = [k.matrix for k in subset]
    if len(gate_mats) == 1:
        grouped_gate_mats.append(gate_mats[0])
    else:
        grouped_gate_mats.append(multi_dot(gate_mats))

mats = []
for i in reversed(range(len(grouped_gate_mats))):
    wire_list = grouped_gates[i][0].wires.tolist()
    gate_mat = grouped_gate_mats[i]
    if len(wire_list) == 1:
        if wire_list[0] == 0:
            mats.append(np.kron(gate_mat, np.eye(2)))
        else:
            mats.append(np.kron(np.eye(2), gate_mat))
    else:
        mats.append(gate_mat)

decomposed_matrix = multi_dot(mats)
assert np.allclose(decomposed_matrix, op.matrix, atol=tol, rtol=0)
```

Reply:
> A small addition to trying out the point Josh raised is to enforce the decomposition by having `dev.operations.remove('CU3')`:
> 
> ```python
> import pennylane as qml
> 
> dev = qml.device("default.qubit", wires=2)
> 
> @qml.qnode(dev)
> def circuit(x, y, z):
>     qml.CU3(x, y, z, wires=[0, 1])
>     return qml.expval(qml.PauliZ(0))
> 
> dev.operations.remove('CU3')
> 
> circuit(0.5, 0.1, 0.2)
> ```
> 
> ```
>     519     def decomposition(theta, phi, lam, wires):
>     520         decomp_ops = [
> --> 521             PhaseShift((lam + phi) / 2, wires=wires[0]),
>     522             PhaseShift((lam - phi) / 2, wires=wires[1]),
>     523             CNOT(wires=wires),
> 
> TypeError: unsupported operand type(s) for +: 'Variable' and 'Variable'
> ```

Now this works as well, @antalszava !